### PR TITLE
xpr updates

### PR DESCRIPTION
**Pull request by spoorendonk**
Updates to the OsiXpr interface needed for coin-or/dip#129

### CODE REVIEW

File: src/OsiXpr/OsiXprSolverInterface.cpp [hunk 1]
```diff
@@ -69,18 +69,24 @@ static void reporterror(const char *fname, int iline, int ierr)
...
-    XPRS_CHECKED(XPRSminim, (prob_, "l"));
+    XPRS_CHECKED(XPRSminim, (prob_, args));
   } else if (objsense_ == -1.0) {
-    XPRS_CHECKED(XPRSmaxim, (prob_, "l"));
+    XPRS_CHECKED(XPRSmaxim, (prob_, args));
   }
 #else
-
-  XPRS_CHECKED(XPRSlpoptimize, (prob_, "l"));
+  XPRS_CHECKED(XPRSmipoptimize, (prob_, args));
```
Comment (svigerske) [inline src/OsiXpr/OsiXprSolverInterface.cpp]: This doesn't look right. `initialSolve()` (and `resolve()`) are for solving an LP, or the LP relaxation of a MIP, only. `branchAndBound()` is for solving a MIP.

Comment (spoorendonk) [reply src/OsiXpr/OsiXprSolverInterface.cpp]: I was a little fast. Should maybe be
```suggestion
  XPRS_CHECKED(XPRSlpoptimize, (prob_, args));
```

and `l` should not be in `args` for LPs. It solves the LP relaxaion only when calling `XPRSmipoptimize` that is why it worked.

Comment (svigerske) [reply src/OsiXpr/OsiXprSolverInterface.cpp]: Probably.

**Comment by svigerske** [comment]
Also a CHANGELOG entry that says more than "updates needed for somewhere else" would be good.

File: CHANGELOG [hunk 2]
```diff
@@ -1,3 +1,8 @@
+16/06/202 OsiXpr
```
Comment (svigerske) [inline CHANGELOG]:
```suggestion
16/06/2020 OsiXpr
```

File: src/OsiXpr/OsiXprSolverInterface.cpp [hunk 3]
```diff
@@ -137,8 +149,15 @@ void OsiXprSolverInterface::branchAndBound()
...
+  const char * args = "";
+  bool takeHint;
+  OsiHintStrength strength;
+  getHintParam(OsiDoDualInResolve, takeHint, strength);
+  if (strength != OsiHintIgnore) {
+    args = takeHint ? "d" : "p";
+  }
 
-  XPRS_CHECKED(XPRSmipoptimize, (prob_, ""));
+  XPRS_CHECKED(XPRSmipoptimize, (prob_, args));
```
Comment (svigerske) [inline src/OsiXpr/OsiXprSolverInterface.cpp:160]: So if one passes 'p' or 'd' here, will that then only solve the LP relaxation, or will Xpress use this argument to decide how to solve the initial LP relaxation in B&B?
If the latter, then taking `OsiDoDualInInitial` may fit better?

Comment (spoorendonk) [reply src/OsiXpr/OsiXprSolverInterface.cpp:160]: As I read the docs https://www.fico.com/fico-xpress-optimization/docs/dms2020-02/solver/optimizer/HTML/XPRSmipoptimize.html then 'p' and 'd' is for the initial LP relaxation. One must pass also the 'l' flag to only solve the LP.

`OsiDoDualInInitial` seems like the right choice

**Comment by tkralphs** [comment]
Umm, shouldn't we wait for the tests to run?

**Comment by svigerske** [comment]
> Umm, shouldn't we wait for the tests to run?

I haven't seen any automated tests for OsiXpr setup by you or the Osi maintainers.

There was a change in OsiXprSolverInterfaceTest.cpp, so I assume that @spoorendonk has run the Osi unittest with Xpress on his machine.

**Comment by spoorendonk** [comment]
> I haven't seen any automated tests for OsiXpr setup by you or the Osi maintainers.

No automatic tests.

> There was a change in OsiXprSolverInterfaceTest.cpp, so I assume that @spoorendonk has run the Osi unittest with Xpress on his machine.

Kind of. It fails for OsiXpr in some cases and I was not able to figure out a solution for all of them. Some because functionality is not implemented and some because the LP solver behaves differently on the problem being solved. I wanted to look into at later point ...

**Comment by tkralphs** [comment]
Well, OK, sorry, I guess we can't run the tests for Xpress automatically. There's probably no point.

**Pull request merged**